### PR TITLE
Change elasticsearch mapping to use CompanyExportCountry

### PR DESCRIPTION
**Pull request by yozzo**
### Description of change

Change elasticsearch mapping to use CompanyExportCountry for export countries data

### Checklist

* [ ] Has a new newsfragment been created? Check [changelog/README.md](https://github.com/uktrade/data-hub-api/blob/master/changelog/README.md) for instructions
* [ ] Do any added or updated endpoints appear in the API documentation? See [docs/Maintaining the API documentation.md](https://github.com/uktrade/data-hub-api/blob/develop/docs/Maintaining&#32;the&#32;API&#32;documentation.md) for more details
* [ ] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [ ] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [ ] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

### CODE REVIEW

File: datahub/search/company/models.py [hunk 1]
```diff
@@ -130,6 +140,14 @@ class Company(BaseESModel):
             'get_one_list_group_global_account_manager',
             dict_utils.contact_or_adviser_dict,
         ),
+        'export_to_countries': lambda obj: [
+            _get_country_id_dict(o) for o in obj.export_countries.all()
```
Comment (reupen) [inline datahub/search/company/models.py]: Is there a reason we've changed from `dict_utils.id_name_dict` to `_get_country_id_dict` here?

File: datahub/search/company/test/test_entity_search_views.py [hunk 2]
```diff
@@ -47,37 +51,59 @@ def setup_data(es_with_collector):
...
-        ],
-        future_interest_countries=[
-            constants.Country.japan.value.id,
-            constants.Country.united_states.value.id,
-        ],
     )
-    CompanyFactory(
+    CompanyExportCountryFactory(
+        company=company1,
+        country=Country.objects.get(pk=constants.Country.france.value.id),
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: Can do something like this instead:

```suggestion
        country_id=constants.Country.france.value.id,
```

(same for the other ones below).

File: datahub/search/company/models.py (hunk 1 above)
Comment (yozzo) [reply datahub/search/company/models.py]: object has no `name` attribute

**Comment by yozzo** [comment]
blocked until https://github.com/uktrade/data-hub-api/pull/2375 is merged

File: datahub/search/company/models.py (hunk 1 above)
Comment (reupen) [reply datahub/search/company/models.py]: Ah – have replied above.

File: datahub/search/company/models.py [hunk 3]
```diff
@@ -22,6 +23,15 @@ def _adviser_field_with_indexed_id():
     )
 
 
+def _get_country_id_dict(obj):
+    if obj is None:
+        return
```
Comment (reupen) [inline datahub/search/company/models.py]: This we can drop this bit – doesn't look like it's possible for `obj` to be `None`.

File: datahub/search/company/models.py [hunk 4]
```diff
@@ -22,6 +23,15 @@ def _adviser_field_with_indexed_id():
...
+
+    return {
+        'id': str(obj.country_id),
```
Comment (reupen) [inline datahub/search/company/models.py]: We're omitting `name` here (which is in the mapping), is there any reason for that? It was there before so seems like we should keep it (and change the query set to prefetch `export_countries__country` instead).

**Comment by yozzo** [comment]
CSV feature currently blocked due to a Django bug: https://code.djangoproject.com/ticket/31097

File: datahub/core/query_utils.py [hunk 5]
```diff
@@ -62,7 +62,7 @@ def __init__(self, **kwargs):
         super().__init__(*args)
 
 
-def get_string_agg_subquery(model, expression, delimiter=', ', distinct=False):
+def get_string_agg_subquery(model, expression, delimiter=', ', distinct=False, filter=None):
```
Comment (reupen) [inline datahub/core/query_utils.py]: Can we remove the `filter` addition here if we aren't using it?

File: datahub/search/company/test/test_entity_search_views.py [hunk 6]
```diff
@@ -887,12 +928,18 @@ def test_export(
...
-                    company.export_to_countries.order_by('name'),
-                ),
-                'Countries of interest': join_attr_values(
-                    company.future_interest_countries.order_by('name'),
-                ),
+                'Countries exported to': ', '.join([
+                    e.country.name for e in list(CompanyExportCountry.objects.filter(
+                        company=company,
+                        status=CompanyExportCountry.EXPORT_INTEREST_STATUSES.currently_exporting,
+                    ).order_by('country__name')) if hasattr(e, 'country')
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: Is ` if hasattr(e, 'country')` needed here?

File: datahub/search/company/test/test_entity_search_views.py [hunk 7]
```diff
@@ -887,12 +928,18 @@ def test_export(
...
                 'Country': get_attr_or_none(company, 'address_country.name'),
                 'UK region': get_attr_or_none(company, 'uk_region.name'),
-                'Countries exported to': join_attr_values(
-                    company.export_to_countries.order_by('name'),
-                ),
-                'Countries of interest': join_attr_values(
-                    company.future_interest_countries.order_by('name'),
-                ),
+                'Countries exported to': ', '.join([
+                    e.country.name for e in list(CompanyExportCountry.objects.filter(
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: Should be able to get rid of the `list()` call here.

Also, should be able to use `company.export_countries.filter(status=...)` instead of `CompanyExportCountry.objects.filter(company=company, status=...)` to be a bit more concise.

File: datahub/search/company/views.py [hunk 8]
```diff
@@ -154,6 +154,8 @@ class PublicSearchCompanyAPIView(HawkResponseSigningMixin, SearchAPIView):
 class SearchCompanyExportAPIView(SearchCompanyAPIViewMixin, SearchExportAPIView):
     """Company search export view."""
 
+    CURRENTLY_EXPORTING = CompanyExportCountry.EXPORT_INTEREST_STATUSES.currently_exporting
+    FUTURE_INTEREST = CompanyExportCountry.EXPORT_INTEREST_STATUSES.future_interest
```
Comment (reupen) [inline datahub/search/company/views.py]: I think it would be better to fix the verbose names where they're defined rather than alias them where they're used.

So e.g. rename `CompanyExportCountry.EXPORT_INTEREST_STATUSES` to `CompanyExportCountry.STATUS` or move `EXPORT_INTEREST_STATUSES` off the class (in a separate PR probably).

File: datahub/core/query_utils.py [hunk 9]
```diff
@@ -77,7 +77,13 @@ def get_string_agg_subquery(model, expression, delimiter=', ', distinct=False):
...
     return get_aggregate_subquery(
         model,
-        StringAgg(expression, delimiter, ordering=(expression,), distinct=distinct),
+        StringAgg(
+            expression,
+            delimiter,
+            ordering=(expression,),
+            distinct=distinct,
+            output_field=CharField(),
+        ),
```
Comment (reupen) [inline datahub/core/query_utils.py]: Is this change still needed?

File: datahub/search/company/test/test_entity_search_views.py [hunk 10]
```diff
@@ -47,37 +50,61 @@ def setup_data(es_with_collector):
     country_us = constants.Country.united_states.value.id
     country_anguilla = constants.Country.anguilla.value.id
     uk_region = constants.UKRegion.south_east.value.id
-    CompanyFactory(
+    company1 = CompanyFactory(
```
Comment (currycoder) [inline datahub/search/company/test/test_entity_search_views.py]: Optional: How about `uk_company` and `us_company` instead of `company_1` and `company_2`?

File: datahub/search/company/test/test_entity_search_views.py [hunk 11]
```diff
@@ -840,24 +867,38 @@ def test_export(
         orm_ordering,
     ):
         """Test export of company search results."""
-        CompanyFactory.create_batch(
+        companies_1 = CompanyFactory.create_batch(
```
Comment (currycoder) [inline datahub/search/company/test/test_entity_search_views.py]: Optional: How about `companies` and `hq_companies` instead of `companies_1` and `companies_2`?
